### PR TITLE
Add decimal support

### PR DESCRIPTION
**Pull request by SpencerC**
Contains #9 
Contains #11

### CODE REVIEW

File: protoconvert.go [hunk 1]
```diff
@@ -69,7 +70,11 @@ func valueToProto(value any) (*pb.Value, error) {
 		return &pb.Value{Inner: &pb.Value_Int{Int: v.UnixMilli()}}, nil
 	case time.Time:
 		return &pb.Value{Inner: &pb.Value_Int{Int: v.UnixMilli()}}, nil
-		// TODO: add decimal support
+	case *big.Float:
+		return &pb.Value{Inner: &pb.Value_Decimal{Decimal: &pb.Decimal{
+			Scale: uint32(v.Prec()),
+			Value: []byte(v.Text('b', int(v.Prec()))),
```
Comment (mpenick) [inline protoconvert.go]: There are a couple issues:
* `decimal` is used to store exact values and floats represent approximate values. `big.Float` might not be the best choice here. Here's some more information on the differences: https://github.com/golang/go/issues/12127. It looks like https://github.com/shopspring/decimal is recommended, but gocql uses https://pkg.go.dev/gopkg.in/inf.v0 (which looks to be dead and experimental). 
* The encoding format is incorrect. Here's some information on how the encoding is represent on the wire: https://github.com/apache/cassandra/blob/cassandra-4.0.5/doc/native_protocol_v3.spec#L804-L807. gocql is a pretty good reference for seeing how to marshal this type along with other DataStax drivers. This encoding seems to be fine if this is the only client, but if you use cqlsh for example you'll see the following for `big.NewFloat(2.78).SetPrec(3))`:

```
cassandra@cqlsh> SELECT decimal_col FROM test.all_types ;

 decimal_col
-------------
  913321.265

(1 rows)
```

Comment (SpencerC) [reply protoconvert.go]: `shopspring/decimal` looks good. How's this?

Comment (mpenick) [reply protoconvert.go]: It looks like the encoding for the varint part is not correct (it looks like it's using Gob encoding). For the value: `decimal.NewFromString("-1.23456789")` I'm seeing the following in `cqlsh`:

```
cassandra@cqlsh> SELECT decimal_col FROM test.all_types ;

 decimal_col
--------------
 130.08358677
```

It looks like the big integer part is private in `shopspring/decimal`. If we could get access that encoding would be pretty straightforward.

Comment (mpenick) [reply protoconvert.go]: This seems to do the trick, using `decimal.Coefficient()` (unfortunately it makes a copy and I'm open to opinions there)  and the `varint` encoding/decoding code:

```diff
diff --git a/protoconvert.go b/protoconvert.go
index 171a534..4ae3a09 100644
--- a/protoconvert.go
+++ b/protoconvert.go
@@ -1,8 +1,8 @@
 package astra
 
 import (
-	"encoding/binary"
 	"fmt"
+	"math/big"
 	"net"
 	"time"
 
@@ -135,14 +135,7 @@ func protoToValue(value *pb.Value) (any, error) {
 	case *pb.Value_Time:
 		return time.Duration(v.Time), nil
 	case *pb.Value_Decimal:
-		d := make([]byte, 4)
-		binary.BigEndian.PutUint32(d, -v.Decimal.Scale)
-		d = append(d, v.Decimal.Value...)
-
-		dec := decimal.New(0, 0)
-		if err := dec.UnmarshalBinary(d); err != nil {
-			return nil, fmt.Errorf("failed to parse decimal: %w", err)
-		}
+		dec := decimal.NewFromBigInt(decodeBigInt(v.Decimal.Value), int32(-v.Decimal.Scale))
 		return dec, nil
 		// TODO: add UDT support
 		// TODO: add varint support
@@ -155,12 +148,39 @@ func protoToValue(value *pb.Value) (any, error) {
 }
 
 func encodeDecimal(d *decimal.Decimal) (*pb.Value, error) {
-	b, err := d.MarshalBinary()
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal decimal binary: %w", err)
-	}
 	return &pb.Value{Inner: &pb.Value_Decimal{Decimal: &pb.Decimal{
 		Scale: uint32(-d.Exponent()),
-		Value: b[4:]},
+		Value: encodeBigInt(d.Coefficient())},
 	}}, nil
 }
+
+func decodeBigInt(data []byte) *big.Int {
+	l := len(data)
+	i := big.NewInt(0).SetBytes(data)
+	if l > 0 && data[0]&0x80 > 0 {
+		i.Sub(i, big.NewInt(0).Lsh(big.NewInt(1), uint(l)*8))
+	}
+	return i
+}
+
+func encodeBigInt(i *big.Int) []byte {
+	switch i.Sign() {
+	case 0:
+		return []byte{0}
+	case 1:
+		b := i.Bytes()
+		if b[0]&0x80 > 0 {
+			b = append([]byte{0}, b...)
+		}
+		return b
+	case -1:
+		l := uint(i.BitLen()/8+1) * 8
+		ii := big.NewInt(0).Add(i, big.NewInt(0).Lsh(big.NewInt(1), l))
+		b := ii.Bytes()
+		if len(b) >= 2 && b[0] == 0xff && b[1]&0x80 != 0 {
+			b = b[1:]
+		}
+		return b
+	}
+	return nil
+}
```

Comment (SpencerC) [reply protoconvert.go]: Changes applied.

**Review by mpenick — APPROVED** [review]
:+1:

**Review by mpenick — APPROVED** [review]

**Pull request merged**